### DISCUSSION TITLE
Getting text content of html element without callback hell

### DISCUSSION THREAD

**Discussion by soadzoor** [cypress-io/cypress#26082] [Questions and Help] [answered]
Hey,

I'm trying to get the text of an html element in some of my cypress test files.
The goal is to check the number of rows in my table, add some elements, then check the number of rows in my table again, if it changed to the right value.

Right now, I'm solving it this way (this is a simplified version):
```
cy.get('.TableFooter > span').invoke('text').then(text => {
    const numberOfRowsBeforeCreation = parseInt(text);

    // ...add items...

    cy.get('.TableFooter > span').invoke('text').then(text => {
        const numberOfRowsAfterCreation = parseInt(text);
        expect(numberOfRowsAfterCreation).to.eq(numberOfRowsBeforeCreation + 1);
    }
});
```

As I said, this example is simplified.. In my real-life test, I'm like 5-6 indentations to the right somewhere in the code because of this callback hell effect...

I also tried making a custom Promise-based function, like..:
```
const getNumberOfRowsInTable = () => {
    return new Promise((resolve, reject) => {
        cy.get('.TableFooter > span').invoke('text').then(text => {
            const numberOfRowsAfterCreation = parseInt(text);
            resolve(numberOfRowsAfterCreation);
        });
    });
};
```

so I was able to use it like this:
```
const numberOfRowsBeforeCreation = await getNumberOfRowsInTable();

// do stuff, add elements, whatever

const numberOfRowsAfterCreation = await getNumberOfRowsInTable();
expect(numberOfRowsAfterCreation).to.eq(numberOfRowsBeforeCreation + 1);
```

It worked fine, but only when I ran the tests with `cypress open`, which opens up the UI, and then I can manually select to run the tests.

If I tried with `npx cypress run`, it got stuck, and failed. I investigated a lot until I found that the Promises are the problem, they're not handled by cypress in this mode correctly, for some weird reason...

Any ideas on how to achieve a nice "line-by-line" structure without actual promises and async await..? Or how to get the text content of an element without the necessity of callback functions...?

Thanks in advance!

**Comment by portyu9** [comment, marked as answer]
Hi @soadzoor 

Don't wrap Cypress commands in a native `Promise` or use `async/await`. Cypress commands already run sequentially through Cypress's command queue.

Make the helper return the Cypress chain instead:

```js
const getNumberOfRows = () =>
  cy.get('.TableFooter > span')
    .invoke('text')
    .then(Number)

let rowsBefore

getNumberOfRows().then((count) => {
  rowsBefore = count
})

// add items here using Cypress commands

getNumberOfRows().then((rowsAfter) => {
  expect(rowsAfter).to.eq(rowsBefore + 1)
})
```

This keeps the callbacks flat—no nesting is required.

The important fix is to **return Cypress chains from helper functions instead of wrapping `cy.*` commands in native Promises**. Cypress commands cannot be `await`ed.

**Reply by portyu9** [reply]
@soadzoor Please mark this answer as accepted if it helped, thank you.

**Reply by soadzoor** [reply]
@portyu9  thank you!
